src/app/qgsgeometryvalidationservice.cpp

```diff
@@ -0,0 +1,454 @@
+  {
+    for ( QMetaObject::Connection connection : qgis::as_const( checkInformation.connections ) )
+    {
+      disconnect( connection );
+    }
+    checkInformation.connections.clear();
+    return;
+  }
+
+  int precision = 0;
+  if ( layer->geometryOptions()->geometryPrecision() == 0 )
+    precision = 8;
+  else
+  {
+    precision = log10( layer->geometryOptions()->geometryPrecision() ) * -1;
+
+    if ( precision < 1 || precision > 13 )
+      precision = 8;
+  }
+
+  checkInformation.context = qgis::make_unique<QgsGeometryCheckContext>( 10, mProject->crs(), mProject->transformContext() );
+
+  QList<QgsGeometryCheck *> layerChecks;
+
+  QgsGeometryCheckRegistry *checkRegistry = QgsAnalysis::instance()->geometryCheckRegistry();
+
+  const QStringList activeChecks = layer->geometryOptions()->geometryChecks();
+
+  const QList<QgsGeometryCheckFactory *> singleCheckFactories = checkRegistry->geometryCheckFactories( layer, QgsGeometryCheck::FeatureNodeCheck, QgsGeometryCheck::AvailableInValidation );
+
+  for ( QgsGeometryCheckFactory *factory : singleCheckFactories )
+  {
+    const QString checkId = factory->id();
+    if ( activeChecks.contains( checkId ) )
+    {
+      const QVariantMap checkConfiguration = layer->geometryOptions()->checkConfiguration( checkId );
+      layerChecks.append( factory->createGeometryCheck( checkInformation.context.get(), checkConfiguration ) );
+    }
+  }
+
+  QList<QgsSingleGeometryCheck *> singleGeometryChecks;
+  for ( QgsGeometryCheck *check : qgis::as_const( layerChecks ) )
+  {
+    Q_ASSERT( dynamic_cast<QgsSingleGeometryCheck *>( check ) );
+    singleGeometryChecks.append( dynamic_cast<QgsSingleGeometryCheck *>( check ) );
+  }
+
+  checkInformation.singleFeatureChecks = singleGeometryChecks;
+
+  // Topology checks
+  QList<QgsGeometryCheck *> topologyChecks;
+  const QList<QgsGeometryCheckFactory *> topologyCheckFactories = checkRegistry->geometryCheckFactories( layer, QgsGeometryCheck::LayerCheck, QgsGeometryCheck::AvailableInValidation );
+
+  for ( QgsGeometryCheckFactory *factory : topologyCheckFactories )
+  {
+    const QString checkId = factory->id();
+    if ( activeChecks.contains( checkId ) )
+    {
+      const QVariantMap checkConfiguration = layer->geometryOptions()->checkConfiguration( checkId );
+      topologyChecks.append( factory->createGeometryCheck( checkInformation.context.get(), checkConfiguration ) );
+    }
+  }
+
+  checkInformation.topologyChecks = topologyChecks;
+
+  // Connect to all modifications on a layer that can introduce a geometry or topology error
+  // Also connect to the beforeCommitChanges signal, so we can trigger topology checks
+  // We keep all connections around in a list, so if in the future all checks get disabled
+  // we can kill those connections to be sure the layer does not even get a tiny bit of overhead.
+  checkInformation.connections
+      << connect( layer, &QgsVectorLayer::featureAdded, this, [this, layer]( QgsFeatureId fid )
```

Comment:
Unique connections do not work for lambdas -- check Qt docs.

Reply:
Makes sense even without reading the docs.
These are here just for safety, it shouldn't normally happen. However I just replaced it with another safety measure to be on the safe side if this is suddenly called in other circumstances in the future.